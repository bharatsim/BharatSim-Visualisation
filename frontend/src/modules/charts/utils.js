import { chartStyleConfig, getColor } from './chartStyleConfig';

function trasformDataForChart(csvData, xColumn, yColumns) {
  return {
    labels: csvData && csvData.data[xColumn],
    datasets:
      csvData &&
      yColumns.map((yColumn, index) => {
        return {
          ...chartStyleConfig,
          label: yColumn,
          borderColor: getColor(index),
          backgroundColor: getColor(index),
          data: csvData && csvData.data[yColumn],
        };
      }),
  };
}

function getYaxisNames(yColumns) {
  return yColumns.map((yColumn) => yColumn.name);
}

export { trasformDataForChart, getYaxisNames };
