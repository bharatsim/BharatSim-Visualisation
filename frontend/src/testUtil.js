// eslint-disable-next-line import/no-extraneous-dependencies
import { fireEvent, within } from '@testing-library/react';

export function selectDropDownOption(container, dropDownId, optionId) {
  const dropDown = container.getByTestId(dropDownId);
  fireEvent.mouseDown(within(dropDown).getByRole('button'));
  const options = within(
    within(document.getElementById(`menu-${dropDownId}`)).getByRole('listbox'),
  );
  fireEvent.click(options.getByTestId(`${dropDownId}-${optionId}`));
}
