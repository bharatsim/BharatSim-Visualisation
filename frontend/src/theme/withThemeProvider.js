import React from 'react';
import { MuiThemeProvider, CssBaseline } from '@material-ui/core';

import theme from './theme';

function withThemeProvider(WrappedComponent) {
  return function WithTheme(props) {
    return (
      <MuiThemeProvider theme={theme}>
        <CssBaseline />
        <WrappedComponent {...props} />
      </MuiThemeProvider>
    );
  };
}

export default withThemeProvider;
