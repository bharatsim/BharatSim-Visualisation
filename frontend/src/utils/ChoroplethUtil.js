import L from 'leaflet';
import chroma from 'chroma-js';

function highlightOnClick(e) {
  const layer = e.target;

  layer.setStyle({
    weight: 2,
    dashArray: '',
    opacity: 1,
    fillOpacity: 0.7,
  });

  if (!L.Browser.ie && !L.Browser.opera && !L.Browser.edge) {
    layer.bringToFront();
  }
}

function resetHighlight(geoJson) {
  return (e) => geoJson.current.resetStyle(e.target);
}

function zoomToFeature(map) {
  return (e) => map.fitBounds(e.target.getBounds());
}

function geoJSONStyle(idDataMap, mapLayerIdName, maxOfMeasure, scale) {
  const colors = Object.keys(scale)
    .sort((v1, v2) => Number(v1) - Number(v2))
    .map((key) => scale[key]);
  const getColor = chroma.scale(colors);
  return (feature) => {
    const gradientPercentage = idDataMap[feature.properties[mapLayerIdName]] / maxOfMeasure;
    return {
      color: '#1f2021',
      weight: 1,
      fillOpacity: 0.5,
      opacity: 0.7,
      fillColor: getColor(gradientPercentage),
    };
  };
}

export { highlightOnClick, resetHighlight, geoJSONStyle, zoomToFeature };
