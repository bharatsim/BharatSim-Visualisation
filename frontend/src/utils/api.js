import { contentTypes, httpMethods } from '../constants/fetch';
import { url as serviceURL } from './url';
import { fetchData, uploadData } from './fetch';

function headerBuilder({ contentType }) {
  return {
    'content-type': contentType,
  };
}

function formDataBuilder(data) {
  const formData = new FormData();
  data.forEach(({ name, value }) => formData.append(name, value));
  return formData;
}

const api = {
  saveDashboard: async ({ widgets, layout, dashboardId, name, count }) => {
    return uploadData({
      url: serviceURL.DASHBOARD_URL,
      headers: headerBuilder({ contentType: contentTypes.JSON }),
      data: JSON.stringify({
        dashboardData: { widgets, layout, dashboardId, name, count },
      }),
    });
  },

  addNewDashboard: async ({ name, projectId }) => {
    return uploadData({
      url: serviceURL.INSERT_DASHBOARD,
      headers: headerBuilder({ contentType: contentTypes.JSON }),
      data: JSON.stringify({
        dashboardData: { widgets: [], layout: [], name, count: 0, projectId },
      }),
      isCustomErrorHandler: true,
    });
  },

  getAllDashBoardByProjectId: async (projectId) =>
    fetchData({
      url: serviceURL.DASHBOARD_URL,
      query: { projectId, columns: ['name', '_id'] },
    }),

  getAllDashBoard: async () => fetchData({ url: serviceURL.DASHBOARD_URL }),

  uploadFileAndSchema: async ({ file, schema, dashboardId }) =>
    uploadData({
      url: serviceURL.DATA_SOURCES,
      data: formDataBuilder([
        { name: 'datafile', value: file },
        { name: 'schema', value: JSON.stringify(schema) },
        { name: 'dashboardId', value: dashboardId },
      ]),
      headers: headerBuilder({ contentType: contentTypes.FILE }),
    }),

  getCsvHeaders: async (dataSourceId) => fetchData({ url: serviceURL.getHeaderUrl(dataSourceId) }),

  getDatasources: async (dashboardId) =>
    fetchData({
      url: serviceURL.DATA_SOURCES,
      query: { dashboardId },
    }),

  getData: async (datasource, columns) =>
    fetchData({ url: serviceURL.getDataUrl(datasource), query: { columns } }),

  saveProject: async ({ id, ...data }) => {
    const requestObject = {
      url: serviceURL.PROJECT_URL,
      headers: headerBuilder({ contentType: contentTypes.JSON }),
      isCustomErrorHandler: true,
    };
    if (id) {
      return uploadData({
        ...requestObject,
        data: JSON.stringify({ projectData: { ...data, id } }),
        method: httpMethods.PUT,
      });
    }
    return uploadData({
      ...requestObject,
      data: JSON.stringify({ projectData: { ...data } }),
      method: httpMethods.POST,
    });
  },

  getProjects: async () => {
    return fetchData({
      url: serviceURL.PROJECT_URL,
    });
  },
  getProject: async (id) => {
    return fetchData({
      url: serviceURL.getProjectUrl(id),
    });
  },
};

export { api };
