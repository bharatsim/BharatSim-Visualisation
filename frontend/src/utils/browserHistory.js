// eslint-disable-next-line import/no-mutable-exports
let BrowserHistory;

function initHistory(history) {
  BrowserHistory = history;
}

function navigateToPreviousPage() {
  BrowserHistory.goBack();
}

function navigateToHome() {
  BrowserHistory.push('/');
}

function reloadPage() {
  // eslint-disable-next-line no-restricted-globals
  location.reload();
}

export { initHistory, navigateToPreviousPage, reloadPage, BrowserHistory, navigateToHome };
