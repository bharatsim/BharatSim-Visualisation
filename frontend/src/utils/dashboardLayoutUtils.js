export function getNewWidgetLayout(numberOfWidgetAdded, cols, count) {
  return createLayout({
    id: `widget-${count}`,
    xPosition: (numberOfWidgetAdded * 6) % cols,
    yPosition: Infinity, // puts it at the bottom
    width: 6,
    height: 2,
  });
}

function createLayout({ id, xPosition, yPosition, width, height }) {
  return {
    i: id,
    x: xPosition,
    y: yPosition,
    w: width,
    h: height,
  };
}
