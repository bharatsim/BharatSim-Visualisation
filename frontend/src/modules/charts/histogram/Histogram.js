import React, { useEffect, useMemo, useState } from 'react';
import PropTypes from 'prop-types';
import Plot from 'react-plotly.js';
import 'react-grid-layout/css/styles.css';
import 'react-resizable/css/styles.css';
import equal from 'fast-deep-equal/es6/react';

import { api } from '../../../utils/api';
import useLoader from '../../../hook/useLoader';
import LoaderOrError from '../../loaderOrError/LoaderOrError';
import { ChartFullSizeWrapper, configs, layoutConfig, tooltip } from '../chartStyleConfig';
import { rgbaToHex } from '../../../utils/helper';

function Histogram({ config }) {
  const { measure, dataSource, color: rgbaColor, axisConfig } = config;
  const [fetchedData, setFetchedData] = useState();
  const { loadingState, message, startLoader, stopLoaderAfterError, stopLoaderAfterSuccess } =
    useLoader();

  async function fetchData() {
    startLoader();
    return api
      .getData(dataSource, [measure])
      .then((data) => {
        stopLoaderAfterSuccess();
        setFetchedData(data);
      })
      .catch(() => {
        stopLoaderAfterError('Unable to fetch data');
      });
  }

  useEffect(() => {
    fetchData();
  }, [measure, dataSource]);

  const onErrorAction = {
    name: 'Retry',
    onClick: fetchData,
  };

  function createData(rawData) {
    const color = rgbaToHex(rgbaColor);
    return [
      {
        x: rawData[measure],
        type: 'histogram',
        name: measure,
        marker: {
          color,
        },
        ...tooltip(measure, color),
      },
    ];
  }

  const chartMemo = useMemo(() => {
    const data = fetchedData && createData(fetchedData.data);
    return { data };
  }, [measure, fetchedData, rgbaColor]);

  const layout = layoutConfig({ xColumn: measure, axisConfig });

  return (
    <LoaderOrError message={message} loadingState={loadingState} errorAction={onErrorAction}>
      <ChartFullSizeWrapper>
        {fetchedData && (
          <Plot
            layout={layout}
            data={chartMemo.data}
            useResizeHandler
            style={{ width: '100%', height: '100%' }}
            config={configs}
          />
        )}
      </ChartFullSizeWrapper>
    </LoaderOrError>
  );
}

Histogram.propTypes = {
  config: PropTypes.shape({
    dataSource: PropTypes.string.isRequired,
    measure: PropTypes.string.isRequired,
    color: PropTypes.shape.isRequired,
    axisConfig: PropTypes.shape({
      xAxisTitle: PropTypes.string,
      yAxisTitle: PropTypes.string,
    }).isRequired,
  }).isRequired,
};

function isEqual(prevProps, newProps) {
  return (
    equal(prevProps.config, newProps.config) &&
    prevProps.layout.h === newProps.layout.h &&
    prevProps.layout.w === newProps.layout.w
  );
}

export default React.memo(Histogram, isEqual);
