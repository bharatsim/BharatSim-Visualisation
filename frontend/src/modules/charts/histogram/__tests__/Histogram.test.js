import { render, waitFor } from '@testing-library/react';
import React from 'react';

import { fireEvent } from '@testing-library/dom';
import { api } from '../../../../utils/api';
import { withRouter } from '../../../../testUtil';
import withThemeProvider from '../../../../theme/withThemeProvider';
import Histogram from '../Histogram';

jest.mock('../../../../utils/api', () => ({
  api: {
    getData: jest.fn().mockResolvedValue({
      data: { exposed: [2, 3], exposedNew: [3, 4] },
    }),
  },
}));

jest.mock('react-plotly.js', () => (props) => (
  <div>
    <span>Histogram</span>
    {/* eslint-disable-next-line no-undef */}
    <pre>{mockPropsCapture(props)}</pre>
  </div>
));

describe('Histogram', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  const HistogramWithProvider = withThemeProvider(withRouter(Histogram));

  it('should create a histogram with selected measure', async () => {
    const { container, findByText } = render(
      <HistogramWithProvider
        config={{
          dataSource: 'dataSource',
          measure: 'exposed',
          color: { r: 1, g: 1, b: 1, a: 1 },
          axisConfig: {
            xAxisTitle: 'xAxisTestTitle',
            yAxisTitle: 'yAxisTestTitle',
          },
        }}
      />,
    );

    await findByText('Histogram');

    expect(container).toMatchSnapshot();
  });

  it('should call get data api for given data column and datasource', async () => {
    const { findByText } = render(
      <HistogramWithProvider
        config={{
          dataSource: 'dataSource',
          measure: 'exposed',
          color: { r: 1, g: 1, b: 1, a: 1 },
          axisConfig: {
            xAxisTitle: 'xAxisTestTitle',
            yAxisTitle: 'yAxisTestTitle',
          },
        }}
      />,
    );

    await findByText('Histogram');

    expect(api.getData).toHaveBeenCalledWith('dataSource', ['exposed']);
  });

  it('should get data whenever column name are updated', async () => {
    const { rerender, findByText } = render(
      <HistogramWithProvider
        config={{
          dataSource: 'dataSource',
          measure: 'exposed',
          color: { r: 1, g: 1, b: 1, a: 1 },
          axisConfig: {
            xAxisTitle: 'xAxisTestTitle',
            yAxisTitle: 'yAxisTestTitle',
          },
        }}
      />,
    );

    await findByText('Histogram');

    rerender(
      <HistogramWithProvider
        config={{
          dataSource: 'dataSource',
          measure: 'exposedNew',
          color: { r: 1, g: 1, b: 1, a: 1 },
          axisConfig: {
            xAxisTitle: 'xAxisTestTitle',
            yAxisTitle: 'yAxisTestTitle',
          },
        }}
      />,
    );

    await findByText('Histogram');

    expect(api.getData).toHaveBeenLastCalledWith('dataSource', ['exposedNew']);
  });

  it('should show loader while fetching data', async () => {
    render(
      <HistogramWithProvider
        config={{
          dataSource: 'dataSource',
          measure: 'exposed',
          color: { r: 1, g: 1, b: 1, a: 1 },
          axisConfig: {
            xAxisTitle: 'xAxisTestTitle',
            yAxisTitle: 'yAxisTestTitle',
          },
        }}
      />,
    );

    const loaderComponent = document.getElementsByTagName('svg');

    expect(loaderComponent).not.toBeNull();

    await waitFor(() => {});
  });

  it('should show error if error occur while fetching data', async () => {
    api.getData.mockRejectedValueOnce('error');
    const { findByText, getByText } = render(
      <HistogramWithProvider
        config={{
          dataSource: 'dataSource',
          measure: 'exposed',
          color: { r: 1, g: 1, b: 1, a: 1 },
          axisConfig: {
            xAxisTitle: 'xAxisTestTitle',
            yAxisTitle: 'yAxisTestTitle',
          },
        }}
      />,
    );

    await findByText('Unable to fetch data');

    expect(getByText('Unable to fetch data')).toBeInTheDocument();
  });

  it('should again fetch data on click on retry button present on error banner', async () => {
    api.getData.mockRejectedValueOnce('error');
    const { findByText, getByText } = render(
      <HistogramWithProvider
        config={{
          dataSource: 'dataSource',
          measure: 'exposed',
          color: { r: 1, g: 1, b: 1, a: 1 },
          axisConfig: {
            xAxisTitle: 'xAxisTestTitle',
            yAxisTitle: 'yAxisTestTitle',
          },
        }}
      />,
    );

    await findByText('Unable to fetch data');
    const retryButton = getByText('Retry').closest('button');
    fireEvent.click(retryButton);

    expect(api.getData).toHaveBeenCalled();
    await waitFor(() => {});
  });
});
