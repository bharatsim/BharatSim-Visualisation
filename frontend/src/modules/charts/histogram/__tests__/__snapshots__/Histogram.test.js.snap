// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Histogram should create a histogram with selected measure 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-4 Styled(MuiBox)-root-3"
    classes="[object Object]"
  >
    <div>
      <span>
        Histogram
      </span>
      <pre>
        layout = {"showlegend":true,"colorway":["#62CEFF","#EC3237","#FAA847","#21C694","#1E3FD8","#C962BE"],"autosize":true,"font":{"family":"Roboto","size":10,"weight":400},"margin":{"l":32,"r":0,"b":60,"t":40,"autoexpand":true},"legend":{"font":{"color":"#929599","family":"Roboto","size":12,"weight":400}},"xaxis":{"title":{"text":"exposed","standoff":8,"font":{"family":"Roboto","size":14,"weight":700}},"color":"#929599","zerolinecolor":"#c5c7cb","linewidth":0.5,"gridwidth":0.5,"zerolinewidth":0.5,"showline":true,"tickwidth":1,"automargin":true,"autotypenumbers":"strict"},"yaxis":{"title":{"standoff":8},"ticklabelposition":"outside","color":"#929599","zerolinecolor":"#c5c7cb","linewidth":0.5,"gridwidth":0.5,"zerolinewidth":0.5,"showline":true,"tickwidth":1,"automargin":true,"autotypenumbers":"strict"},"shapes":[],"annotations":[]}
        data = [{"x":[2,3],"type":"histogram","name":"exposed","hoverlabel":{"bgcolor":"#fff","bordercolor":"#62CEFF","font":{"color":"#000"}},"hoverinfo":"x+y","hovertemplate":"exposed: %{y}&lt;extra&gt;&lt;/extra&gt;"}]
        useResizeHandler = true
        style = {"width":"100%","height":"100%"}
        config = {"modeBarButtonsToRemove":["select2d","lasso2d","toImage"],"displaylogo":false,"responsive":true}
      </pre>
    </div>
  </div>
</div>
`;
